src/crimson/osd/pg.cc

```diff
@@ -1454,14 +1454,19 @@ void PG::on_change(ceph::os::Transaction &t) {
 }
 
 void PG::context_registry_on_change() {
-    obc_registry.for_each([](ObjectContextRef obc) {
-      assert(obc);
-      for (auto j = obc->watchers.begin();
-           j != obc->watchers.end();
-           j = obc->watchers.erase(j)) {
-        j->second->discard_state();
-      }
+  std::vector<seastar::shared_ptr<crimson::osd::Watch>> watchers;
+  obc_registry.for_each([&watchers](ObjectContextRef obc) {
+    assert(obc);
+    for (auto j = obc->watchers.begin();
+         j != obc->watchers.end();
+         j = obc->watchers.erase(j)) {
+      watchers.emplace_back(j->second);
+    }
   });
+
+  for (auto &watcher : watchers) {
+    watcher->discard_state();
+  }
```

Comment:
IIUC we will still have an issue with the scenario mentioned (with the changes applied):

1. a watcher is added to `watchers`
2. the corresponding object context's use_count drops to zero
3. the object context is unreferenced by obc lru
4. obc is deleted from obc lru by lru's evict()
5. discard the watcher's state
6. watcher's discard is corrupted (`ceph_assert(obc)` will fail)

We should check if the watcher's obc wasn't deleted before discarding its state, right?

Reply:
Um..., It looks to me that obc is deleted after all of its watchers are destroyed, so this shouldn't be a problme, am I missing anything?

Reply:
> Um..., It looks to me that obc is deleted after all of its watchers are destroyed, so this shouldn't be a problme

ack